Ginkgo - Add service recovery on restart test

eloycoto opened the pull request:
- Ported `tests/06-lb.sh:test_svc_restore_functionality` test to ginkgo
- Added `cilium.ServiceList` and `cilium.ServiceGetIds` helpers
functions.

Fix #2049

Signed-off-by: Eloy Coto <eloy.coto@gmail.com>

eloycoto committed:
Ginkgo - Add service recovery on restart test

- Ported `tests/06-lb.sh:test_svc_restore_functionality` test to ginkgo
- Added `cilium.ServiceList` and `cilium.ServiceGetIds` helpers
functions.

Fix #2049

Signed-off-by: Eloy Coto <eloy.coto@gmail.com>

ianvernon commented on test/helpers/cilium.go:
```diff
@@ -343,14 +343,30 @@ func (c *Cilium) ServiceAdd(id int, frontend string, backends []string, rev int)
-//ServiceDel delete the service ID
-func (c *Cilium) ServiceDel(id int) *CmdRes {
-	return c.Exec(fmt.Sprintf("service delete '%d'", id))
+// ServiceGetIds it returns and array with all services ids. If can not get the
```
Incorrect grammar; "...returns an array with the IDs of all Cilium services. Returns an error if the IDs cannot be retrieved."

ianvernon commented on test/helpers/cilium.go:
```diff
@@ -343,14 +343,30 @@ func (c *Cilium) ServiceAdd(id int, frontend string, backends []string, rev int)
+	return c.Exec(fmt.Sprintf("service delete '%d'", id))
+}
+
+// ServiceList it returns the `cilium service list` output
```
Incorrect grammar: "... returns the output of `cilium service list`"

ianvernon commented on test/runtime/lb.go:
```diff
@@ -210,15 +210,57 @@ var _ = Describe("RuntimeLB", func() {
+			"client",
+			"curl -s --fail --connect-timeout 4 http://2.2.2.2:80/public")
+		status.ExpectSuccess("LB is not working after restart")
+	})
```
This does not fully port the functionality of `tests/06-lb.sh:test_svc_restore_functionality`. We need to be sure that the services and their associated metadata, not just the IDs, are equal. 
This includes frontends, backends, and IDs. For instance:
```
vagrant@cilium-master:~/go/src/github.com/cilium/cilium$ cilium service list -o json | python -m json.tool
[
    {
        "BackendAddresses": [
            "1 => 10.11.20.8:80",
            "2 => 10.11.94.160:80"
        ],
        "FrontendAddress": "2.2.2.2:80",
        "ID": 1
    }
]
```

Check that for the given ID that its associated FrontendAddress and BackendAddress are equal. Furthermore, there are no checks that the BPF LB maps are in sync which is done in `tests/06-lb.sh`. Please add that functionality.

ianvernon replied:
This still does not fully port the functionality in `tests/06-lb.sh:test_svc_restore_functionality`. It's necessary to check that the BPF maps match the Cilium service maps:
```
local SVC_IDS=$(cilium service list | tail -n+2 | awk '{ print $1 }')
  for id in $SVC_IDS; do
    check_cilium_bpf_maps_synced $id
  done
```

ianvernon commented on test/helpers/cilium.go:
```diff
@@ -343,14 +343,30 @@ func (c *Cilium) ServiceAdd(id int, frontend string, backends []string, rev int)
 	return c.Exec(cmd)
 }
 
+//ServiceDel delete the service ID
```
Incorrect grammar: delete --> deletes. 
Rephrase to say "deletes the service with the specified ID"

ianvernon commented on test/runtime/lb.go:
```diff
@@ -210,15 +210,71 @@ var _ = Describe("RuntimeLB", func() {
+		oldSvcIds, err := cilium.ServiceGetIds()
+		Expect(err).Should(BeNil())
+
+		oldbpflb := cilium.Exec("bpf lb list -o json")
```
Make the variable name easier to read with CamelCase --> oldbpflb --> oldBpfLb

ianvernon replied:
This command is hardcoded twice in this function. Factor out `bpf lb list -o json` into a separate string to reduce the amount of hardcoding.

ianvernon commented on test/runtime/lb.go:
```diff
@@ -210,15 +210,71 @@ var _ = Describe("RuntimeLB", func() {
+
+		svcIds, err := cilium.ServiceGetIds()
+		Expect(err).Should(BeNil())
+		Expect(len(svcIds)).Should(Equal(len(oldSvcIds)), "Service recovery does not match")
```
this error message isn't evocative of what is being checked; specify that the number of services added to Cilium differs before and after restart. Also specify what each value is (num of services before / after restart)

eloycoto replied:
The values are going to be provided by gomega assert

ianvernon replied:
I was wrong, you are correct.

ianvernon commented on test/runtime/lb.go:
```diff
@@ -210,15 +210,71 @@ var _ = Describe("RuntimeLB", func() {
+
+		newSvc := cilium.ServiceList()
+		newSvc.ExpectSuccess("Cannot retrieve service list after restart")
+		newSvc.ExpectEqual(oldSvc.Output().String(), "Service list does not match")
```
There is not enough output for debugging this test failure. Please log the output of `oldSvc.Output.String()` and `newSvc.Output.String()` to make it easier to debug the tests if they fail. Do this also for the output of `bpf lb list -o json` as well.

eloycoto replied:
Hi, 

When you test on an assert, the Ginkgo will report both values and the assert message. So it'll be covered.

ianvernon replied:
That is correct! I was wrong.

ianvernon commented on test/runtime/lb.go:
```diff
@@ -210,15 +210,71 @@ var _ = Describe("RuntimeLB", func() {
+		httpd2, err := docker.ContainerInspectNet("httpd2")
+		Expect(err).Should(BeNil())
+
+		status := cilium.ServiceAdd(1, "2.2.2.2:80", []string{
```
This IP address is hardcoded and used multiple times in this test. Factor it out into a variable, along with the port.

ianvernon reviewed (changes requested):
This does still not fully port the functionality from the bash script test. I also have concerns about liberal hardcoding of identical strings and lack of informative error messages.

eloycoto committed:
Ginkgo- Added ServiceIsSynced method

- Checks that a service is synced in the bpf lb list
- Fix issue with json output on `bpf lb list`

Signed-off-by: Eloy Coto <eloy.coto@gmail.com>

ianvernon commented on test/helpers/cilium.go:
```diff
@@ -343,14 +359,74 @@ func (c *Cilium) ServiceAdd(id int, frontend string, backends []string, rev int)
+	return c.Exec(fmt.Sprintf("service delete '%d'", id))
+}
+
+// ServiceIsSync checks that the service is sync in the bpf lb list
```
Rename to "ServiceIsSynced".

ianvernon replied:
Incorrect grammar: "sync --> "is in sync with"

ianvernon replied:
checks that the Cilium service with the specified id has its metadata match that of the loadbalancer BPF maps

ianvernon commented on test/helpers/cilium.go:
```diff
@@ -343,14 +359,74 @@ func (c *Cilium) ServiceAdd(id int, frontend string, backends []string, rev int)
+		fmt.Sprintf("%d", svc.FrontendAddress.Port))
+	lb, ok := bpfLB[frontendAddr]
+	if ok == false {
+		return false, fmt.Errorf("service is not sync with BPF lb")
```
Specify how it is not in sync (frontend address from the service with the specified ID does not have a corresponding frontend address in the bpf maps, etc.) along with which service being analyzed resulted in this error. Otherwise, in logs, we won't be able to determine which service resulted in this error unless the caller of this function does so.

ianvernon commented on test/helpers/cilium.go:
```diff
@@ -52,6 +53,21 @@ func CreateCilium(target string, log *log.Entry) *Cilium {
 	}
 }
 
+// BpfLBList returns the output of `cilium bpf lb list -o json` as a map
```
Specify what this mapping represents - what do the key / value represent?

ianvernon commented on test/helpers/cilium.go:
```diff
@@ -343,14 +359,74 @@ func (c *Cilium) ServiceAdd(id int, frontend string, backends []string, rev int)
+			}
+		}
+		if result == false {
+			return false, fmt.Errorf("backend addr %s does not match with LB", target)
```
addr --> address

ianvernon replied:
"does not match" is unclear. specify that the backend address does not exist in the BPF maps for the specified frontend and provide context for which service this has occurred.

ianvernon commented on test/helpers/cmd.go:
```diff
@@ -138,6 +138,14 @@ func (res *CmdRes) Output() *bytes.Buffer {
 	return res.stdout
 }
 
+//ExpectEqual makes an assertion that compare cmdRes.Output().String() and the
```
"makes an assertion that compare" --> "asserts whether"

ianvernon commented on test/helpers/cmd.go:
```diff
@@ -138,6 +138,14 @@ func (res *CmdRes) Output() *bytes.Buffer {
 }
 
+//ExpectEqual makes an assertion that compare cmdRes.Output().String() and the
+//given parameter.  It accepts an optional parameter that can be used to
```
Incorrect grammar - "asserts whether .... and expected are equal"
